include/ccf/tx.h

```diff
@@ -0,0 +1,305 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/tx_id.h"
+#include "crypto/hash.h"
+#include "ds/ccf_assert.h"
+#include "kv/kv_types.h"
+#include "kv/untyped_map.h"
+
+#include <list>
+#include <map>
+#include <memory>
+#include <optional>
+#include <string>
+
+namespace kv
+{
+  class CompactedVersionConflict
+  {
+  private:
+    std::string msg;
+
+  public:
+    CompactedVersionConflict(const std::string& s) : msg(s) {}
+
+    char const* what() const
+    {
+      return msg.c_str();
+    }
+  };
+
+  struct MapChanges
+  {
+    // Shared ownership over source map
+    std::shared_ptr<AbstractMap> map;
+
+    // Owning pointer of ChangeSet over that map
+    std::unique_ptr<untyped::ChangeSet> changeset;
+  };
+
+  // When a collection of Maps are locked, the locks must be acquired in a
+  // stable order to avoid deadlocks. This ordered map will claim in name-order
+  using OrderedChanges = std::map<std::string, MapChanges>;
+
+  // Manages a collection of MapHandles. Derived implementations should call
+  // get_handle_by_name to retrieve handles over their desired maps.
+  class BaseTx
+  {
+  protected:
+    AbstractStore* store;
+
+    OrderedChanges all_changes;
+
+    // NB: This exists only to maintain the old API, where this Tx stores
+    // MapHandles and returns raw pointers to them. It could be removed entirely
+    // with a near-identical API if we return `shared_ptr`s, and assuming that
+    // we don't actually care about returning exactly the same Handle instance
+    // if `rw` is called multiple times
+    using PossibleHandles = std::list<std::unique_ptr<AbstractMapHandle>>;
+    std::map<std::string, PossibleHandles> all_handles;
+
+    // In most places we use NoVersion to indicate an invalid version. In this
+    // case, NoVersion is a valid value - it is the version that the first
+    // transaction in the service will read from, before anything has been
+    // applied to the KV. So we need an additional special value to distinguish
+    // "haven't yet fetched a read_version" from "have fetched a read_version,
+    // and it is NoVersion", and we get that by wrapping this in a
+    // std::optional with nullopt representing "not yet fetched".
+    std::optional<Version> read_version = std::nullopt;
+    ccf::View view = ccf::VIEW_UNKNOWN;
```

Comment:
I renamed `Tx.term` to `Tx.view` here, because it made sense at the time (I thought this wouldn't see `Term` any more), but that's an incomplete migration. I'm planning to do a broader rename of these (the variables, types, and getters on `Tx`) to `View` and `SeqNo`, but this PR shows its not part of the public API so its not urgent and can be done later.